### PR TITLE
LogFactory - AutoShutdown to allow users to control shutdown logic

### PR DESCRIPTION
**Pull request by snakefoot**
Support Microsoft Extension IHostBuilder controlling when to shutdown. Because AppDomain-Unload is called premature (compared to IIS).

See also  #3284

Note merge to master

### CODE REVIEW

File: src/NLog/LogFactory.cs [hunk 1]
```diff
@@ -185,6 +185,28 @@ public static IAppDomain CurrentAppDomain
...
+            get { return _autoShutdown; }
+            set
+            {
+                if (value != _autoShutdown)
+                {
+                    _autoShutdown = value;
+#if !SILVERLIGHT && !__IOS__ && !__ANDROID__ && !NETSTANDARD1_3
+                    LoggerShutdown -= OnStopLogging;
+                    if (value)
+                        LoggerShutdown += OnStopLogging;
```
Comment (304NotModified) [inline src/NLog/LogFactory.cs]: @snakefoot should we assert this in the tests? I think this is the most important lines to assert:

```
                    LoggerShutdown -= OnStopLogging;
                    if (value)
                        LoggerShutdown += OnStopLogging;
```

**Comment by snakefoot** [comment]
Didnt want to add test before the initial change was approved. Going to abuse medium trust test of appdomain.


But you approve the direction?


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@304NotModified commented on this pull request.

________________________________

In src/NLog/LogFactory.cs<https://github.com/NLog/NLog/pull/3578#discussion_r320027418>:

> @@ -185,6 +185,25 @@ public static IAppDomain CurrentAppDomain
         /// </summary>
         public bool KeepVariablesOnReload { get; set; }

+        /// <summary>
+        /// Gets or sets a value indicating whether to automatically call <see cref="LogFactory.Shutdown"/>
+        /// on AppDomain.Unload or AppDomain.ProcessExit
+        /// </summary>
+        public bool? AutoShutdown
+        {
+            get { return _autoShutdown; }
+            set
+            {
+#if !SILVERLIGHT && !__IOS__ && !__ANDROID__ && !NETSTANDARD1_3
+                LoggerShutdown -= OnStopLogging;


Add test?

________________________________

In src/NLog/LogFactory.cs<https://github.com/NLog/NLog/pull/3578#discussion_r320489113>:

> +        /// <summary>
+        /// Gets or sets a value indicating whether to automatically call <see cref="LogFactory.Shutdown"/>
+        /// on AppDomain.Unload or AppDomain.ProcessExit
+        /// </summary>
+        public bool AutoShutdown
+        {
+            get { return _autoShutdown; }
+            set
+            {
+                if (value != _autoShutdown)
+                {
+                    _autoShutdown = value;
+#if !SILVERLIGHT && !__IOS__ && !__ANDROID__ && !NETSTANDARD1_3
+                    LoggerShutdown -= OnStopLogging;
+                    if (value)
+                        LoggerShutdown += OnStopLogging;


@snakefoot<https://github.com/snakefoot> should we assert this in the tests? I think this is the most important lines to assert:

                    LoggerShutdown -= OnStopLogging;
                    if (value)
                        LoggerShutdown += OnStopLogging;

**Comment by 304NotModified** [comment]
That's understandable :angel:



> But you approve the direction?

Yes! Looks good, also nice it could be set in the nlog.config 👍 👍

File: src/NLog/LogFactory.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/LogFactory.cs]: Did not see such tests for https://github.com/NLog/NLog.Extensions.Logging/pull/314 :)

But I have now created a test, that fails when autoshutdown=false is not working.

**Pull request merged**